### PR TITLE
CI: Skip tests

### LINKED ISSUE #143: Skip tests on documentation and example changes
Currently we run all our tests every time we open a PR, we should skip running that if only documentation change was made or one of the examples was changed, etc.

### PR DESCRIPTION
**Pull request by kosyfrances**
Currently we run all our tests every time we open a PR,
we should skip running that if only documentation change was made,
or one of the examples was changed.

Fixes #143

### CODE REVIEW

File: changes [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+set -euo pipefail
+
+DOC_EXAMPLE_CHANGE_PATTERN="\
+  -e ^doc/ \
+  -e ^examples/ \
+  -e ^(VERSION|LICENSE)$ \
+  -e \\.md$\
+"
+
+REMOTE="origin/master"
```
Comment (asymmetric) [inline changes]: This is not a remote (`origin` is the remote). Maybe just use the same name as rkt, `BRANCHING_POINT`.

File: changes [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (asymmetric) [inline changes]: Could you rename this to something like `check-changes.sh`?

Also, could you move it to `hacks/`?

File: changes [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  -e ^(VERSION|LICENSE)$ \
+  -e \\.md$\
+"
+
+REMOTE="origin/master"
+
+CHANGES=$(git --no-pager diff --name-only HEAD.."$REMOTE" | grep -vcE "$DOC_EXAMPLE_CHANGE_PATTERN") || true
+
+if [[ "${CHANGES}" -eq 0 ]]; then
+  if [[ "{TRAVIS_PULL_REQUEST}"]]; then
```
Comment (asymmetric) [inline changes]: There is an error here.

Tip: Run your scripts through [shellcheck](https://www.shellcheck.net/) (you can also configure your editor to run it locally)

File: changes [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  -e \\.md$\
+"
+
+REMOTE="origin/master"
+
+CHANGES=$(git --no-pager diff --name-only HEAD.."$REMOTE" | grep -vcE "$DOC_EXAMPLE_CHANGE_PATTERN") || true
+
+if [[ "${CHANGES}" -eq 0 ]]; then
+  if [[ "{TRAVIS_PULL_REQUEST}" != false ]]; then
+    travis_terminate 0
```
Comment (asymmetric) [inline changes]: Wouldn't it be enough to exit with a non-zero value? What do these functions contribute?

Comment (kosyfrances) [reply changes]: According to the docs, https://docs.travis-ci.com/user/customizing-the-build#Breaking-the-Build
```If before_install, install or before_script returns a non-zero exit code, the build is errored and stops immediately``` But if we return a zero exit code, I am not sure the build will stop immediately.

File: .travis.yml [hunk 5]
```diff
@@ -22,6 +22,8 @@ git:
   depth: 1
 
 before_script:
+# Check for changes in documentation and example
+- ./hack/check-changes.sh
```
Comment (asymmetric) [inline .travis.yml]: Seems like the commands below were executed (`curl` etc..) so it would make sense to isolate this script into its own step, so we don't run the rest of the commands if we don't have to.

File: .circleci/config.yml [hunk 6]
```diff
@@ -12,6 +12,8 @@ jobs:
             MINIKUBE_VERSION: v0.26.1
             CHANGE_MINIKUBE_NONE_USER: true
           command: |
+            # Check for changes in documentation and example
+            ./hack/check-changes.sh
```
Comment (asymmetric) [inline .circleci/config.yml]: Seems like the commands below [were executed](https://circleci.com/gh/kinvolk/habitat-operator/175) (`curl` etc..) so it would make sense to isolate this script into its own step, so we don't run the rest of the commands if we don't have to.

**Comment by kosyfrances** [comment]
Here is the log for the build with skipped tests https://circleci.com/gh/kinvolk/habitat-operator/182

File: .circleci/config.yml [hunk 7]
```diff
@@ -5,6 +5,10 @@ jobs:
     working_directory: ~/.go_workspace/src/github.com/habitat-sh/habitat-operator
     steps:
       - checkout
+      - run:
+          name: Check for changes in documentation and example
```
Comment (asymmetric) [inline .circleci/config.yml]: *examples

File: hack/check-changes.sh [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+# This script checks for changes only in documentation and example
+# in order to prevents tests from being run where it is used
+# The idea is to skip tests on documentation and example changes
+# https://github.com/habitat-sh/habitat-operator/issues/143
+
+set -eu
+
+# Finds the branching point of two commits.
+# For example, let B and D be two commits, and their ancestry graph as A -> B, A -> C -> D.
+# Given commits B and D, it returns A.
```
Comment (asymmetric) [inline hack/check-changes.sh]: Could you add a link to the code where we got this from?

With line number, from a specific commit.

File: .circleci/config.yml (hunk 7 above)
Comment (kosyfrances) [reply .circleci/config.yml]: fixed

File: hack/check-changes.sh [hunk 9]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# This script checks for changes only in documentation and examples
+# in order to prevents tests from being run where it is used
+# The idea is to skip tests on documentation and example changes
+# https://github.com/habitat-sh/habitat-operator/issues/143
```
Comment (asymmetric) [inline hack/check-changes.sh]: I would reword this to something like: "This script checks to see if changes only affect documentation and examples, and skips running tests if that's the case".

I also don't think we need a link to that issue.

WDYT?

File: hack/check-changes.sh [hunk 10]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (asymmetric) [inline hack/check-changes.sh]: I guess we could make the name even more descriptive, like `check-skippable-changes`.

File: hack/check-changes.sh [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+function getBranchingPoint {
+    diff --old-line-format='' --new-line-format='' \
+        <(git rev-list --first-parent "${1:-$1}") \
+            <(git rev-list --first-parent "${2:-$2}") | head -1
+}
+
+DOC_EXAMPLE_CHANGE_PATTERN="\
+            -e ^doc/ \
+            -e ^examples/ \
+            -e ^(VERSION|LICENSE)$ \
```
Comment (asymmetric) [inline hack/check-changes.sh]: * `MAINTAINERS`
* `README`
* `CHANGELOG`
* `CODE_OF_CONDUCT`
* `CONTRIBUTING`

File: hack/check-changes.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            -e \\.md$\
+"
+
+BRANCHING_POINT=$(getBranchingPoint HEAD origin/master)
+CHANGES=$(git diff-tree --no-commit-id --name-only -r HEAD.."${BRANCHING_POINT}" | grep -cEv "${DOC_EXAMPLE_CHANGE_PATTERN}") || true
```
Comment (asymmetric) [inline hack/check-changes.sh]: I don't think this is enough. This only tells you if there have been any changes to the docs or examples, but it doesn't tell you if they're the only changes.

If you look at [`rkt`](https://github.com/rkt/rkt/blob/2de014aa1a8f1d9e92bc869e7d666679f2e45a1d/tests/build-and-run-tests.sh#L47), they also run the same `grep` with `-v`, so that they see if there are any changes to the actual source.

Comment (asymmetric) [inline hack/check-changes.sh]: Could you rename this to `SRC_CHANGES`?

File: hack/check-changes.sh (hunk 11 above)
Comment (asymmetric) [reply hack/check-changes.sh]: :man_facepalming: sorry, they're all caught by the `.md`.

File: hack/check-changes.sh (hunk 9 above)
Comment (kosyfrances) [reply hack/check-changes.sh]: I think rewording it is a good idea. Thanks

File: hack/check-skippable-changes.sh [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+# This script checks to see if changes only affect documentation and examples
+# and skips running tests if that's the case
+
+set -eu
+
+# Finds the branching point of two commits.
+# For example, let B and D be two commits, and their ancestry graph as A -> B, A -> C -> D.
+# Given commits B and D, it returns A.
+# https://github.com/rkt/rkt/blob/2de014aa1a8f1d9e92bc869e7d666679f2e45a1d/tests/build-and-run-tests.sh#L31-L38
+function getBranchingPoint {
```
Comment (asymmetric) [inline hack/check-skippable-changes.sh]: Do we need this in practice? Our PRs only merge into master, so we could simply hard-code it.

Thoughts?

/cc @krnowak @iaguis

Comment (asymmetric) [reply hack/check-skippable-changes.sh]: The advantage is that we make the script easier to understand and modify.

The downside is that if we start merging PRs to other branches, we might see breakage. But I don't think that's happening.

Comment (iaguis) [reply hack/check-skippable-changes.sh]: This was done in rkt because it'd compare the PR branch with `origin/master`, which could have changed (other PRs could've gotten merged) and then, if somebody force-pushes to the PR branch, it will rebuild it even if there was only doc changes.

Comment (kosyfrances) [reply hack/check-skippable-changes.sh]: @asymmetric are we good with this keeping the branching point function?

File: .circleci/config.yml [hunk 14]
```diff
@@ -5,6 +5,10 @@ jobs:
     working_directory: ~/.go_workspace/src/github.com/habitat-sh/habitat-operator
     steps:
       - checkout
+      - run:
+          name: Check for changes in documentation and examples
+          command: |
+            ./hack/check-skippable-changes.sh
```
Comment (asymmetric) [inline .circleci/config.yml]: You don't need the `|` here since it's just one line ;)

Comment (kosyfrances) [reply .circleci/config.yml]: fixed :)

File: hack/check-skippable-changes.sh [hunk 15]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+# This script checks to see if changes only affect documentation and examples
+# and skips running tests if that's the case
+
+set -eu
```
Comment (indradhanush) [inline hack/check-skippable-changes.sh]: `set -euo pipefail` is better than `set -eu` as it will make the script fail in scenarios like:
```
foo | tail -f
```

If there is an error in executing `foo`. The current flags will let it pass as `tail -f` would succeed even if `foo` has failed.

File: hack/check-skippable-changes.sh [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+DOC_EXAMPLE_CHANGE_PATTERN="\
+            -e ^doc/ \
+            -e ^examples/ \
+            -e ^(VERSION|LICENSE)$ \
+            -e \\.md$\
+"
+
+BRANCHING_POINT=$(getBranchingPoint HEAD origin/master)
+SRC_CHANGES=$(git diff-tree --no-commit-id --name-only -r HEAD.."${BRANCHING_POINT}" | grep -cEv "${DOC_EXAMPLE_CHANGE_PATTERN}") || true
```
Comment (krnowak) [inline hack/check-skippable-changes.sh]: This won't work.

Passing `${DOC_EXAMPLE_CHANGE_PATTERN}` wrapped in double quotes makes it a single parameter to grep. So grep will try to count how many lines match ` ^doc/ -e ^examples/ -e …`, where `-e`s is a part of the regexp.

Try it yourself (note that I dropped `c` parameter from grep):

```
DOC_EXAMPLE_CHANGE_PATTERN="\
            -e ^doc/ \
            -e ^examples/ \
            -e ^(VERSION|LICENSE)$ \
            -e \\.md$\
"
echo -e "doc/foo\nexamples/yadda\nVERSION\nLICENSE\nwhatever.md\nsrcfile.go" | grep -Ev "${DOC_EXAMPLE_CHANGE_PATTERN}"
```

I'd expect only `srcfile.go` to be printed, but I get:

```
doc/foo
examples/yadda
VERSION
LICENSE
whatever.md
srcfile.go
```

So, either:
- drop the double quotes wrapping the `DOC_EXAMPLE_CHANGE_PATTERN` variable in grep invocation. It feels icky, but makes it work
- use bash array variable and use `"${DOC_EXAMPLE_CHANGE_PATTERN[@]}"` in grep invocation:

```
DOC_EXAMPLE_CHANGE_PATTERN=(
  '-e' '^doc/'
  '-e' '^examples/'
  …
)
```

Note that I'm using single quotes - wrapping regexps in double quotes is just asking for some unwanted variable interpolation.

File: hack/check-skippable-changes.sh [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+DOC_EXAMPLE_CHANGE_PATTERN="\
+            -e ^doc/ \
+            -e ^examples/ \
+            -e ^(VERSION|LICENSE)$ \
+            -e \\.md$\
+"
+
+BRANCHING_POINT=$(getBranchingPoint HEAD origin/master)
```
Comment (kosyfrances) [inline hack/check-skippable-changes.sh]: @indradhanush Using the pipefail option causes this line to return nothing. I am not sure why it happens that way though.

Comment (krnowak) [reply hack/check-skippable-changes.sh]: I guess it is because `diff` returns 1 if there are differences. That triggers the pipefail setting to make the whole expression to return 1, which in turn triggers `set -e` which makes the script to bail out I think.

File: hack/check-skippable-changes.sh (hunk 15 above)
Comment (krnowak) [reply hack/check-skippable-changes.sh]: `pipefail` may be annoying in case of `diff` - `diff` "fails" if the two inputs differ. Would need some workaround for it like, dunno, `(diff --old-line-format='' --new-line-format=''  <(…) <(…) || true) | head -1` I suppose.

File: hack/check-skippable-changes.sh (hunk 17 above)
Comment (indradhanush) [reply hack/check-skippable-changes.sh]: @kosyfrances I see. Ignore my suggestion in that case. :smile:
